keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,447 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Proposal to add a new Kubelet gRPC API with endpoint returning local Pods readiness information.
+Serving Pods statuses by Kubelet within a Node will increase reliability and reduce load to the Kubernetes API Server and traffic outside the node. A connectivity issue between Node and Control Plane should not impact workloads which depend on Pods readiness statuses.
+This API is open to future extension but added information should be limited to pod's readiness information.
+
+## Motivation
+
+Kubelet is responsible for running Health Checks (probes) and communicating the
+results via Pod status. All that information is stored in cache and reported to
+Kube-API. Right now pod's readiness information is tightly coupled with the Kubernetes API
+Server. When a workload wants to know the actual state of Pods running on the
+Node, it needs to fetch it from Kube-API. This causes some issues:
+
+* Reliability - for various reasons Kube-API might not be available
+  (connectivity issue, control plane updates)
+* Scalability - adding new watchers to kube-API is a scalability concern. By
+  exposing the endpoint that will serve the Pods readiness status directly from the Kubelet
+  cache we can use  it on node workloads and avoid additional dependencies on the
+  Kubernetes API Server.
+
+| Impact | Description|
+| ------- | ------------ |
+| + | Reliability - for various reasons kube-API might not be available but this doesn’t mean that local workloads are not accessible and on node system workloads should have the most recent data about pod's readiness even when kube-API is unreachable. |
+| + | Scalability - Reduce the load on kube-API by reducing the number of watchers and using Kubelet to fetch local Pods readiness. Fetching only Pods limited to one node is costly operation for kube-API. |
+| + | Safety - Read-only API will not add security risks. |
+| + | Reduce resource consumption by workload. Using Kube-API we can fetch objects like PodSpec or PodStatus, for some on-node workload this might be unnecessary, with this API workload can reduce the resource consumption. |
+| - | This API will add load to Kubelet (mitigation: API will be rate limited) |
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is designed to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
+
+### Goals
+
+The goal of this API is to expose Pod readiness information directly from the
+source - Kubelet, independent of Control Plane availability by returning kubelet
+owned conditions that are set for pod. This would remove the need for node-local
+components to request this node-local information from
+the Kubernetes API Server.
```

Comment:
My suggestion would be to change the first sentence slightly:

The goal of this API is to expose Pod readiness information directly from the
source - Kubelet, independent of Control Plane availability ~~by returning kubelet
owned conditions that are set for pod~~. This would remove the need for node-local
components to request this node-local information from
the Kubernetes API Server.

I think that proposed removed part doesn't provide the information needed, and again is confusing wrt conditions.

Reply:
done